sql/table.cc

```diff
@@ -6069,7 +6069,7 @@ bool TABLE::add_tmp_key(Field_map *key_parts, bool invisible,
     s->max_key_length = std::max(s->max_key_length, key_len);
     s->key_parts += key_part_count;
     assert(s->keys < s->max_tmp_keys);
-    sprintf(s->key_names[s->keys].name, "<auto_key%d>", s->keys);
+    sprintf(s->key_names[s->keys].name, "<auto_key%d>", s->auto_key_name_server++);
```

Comment:
better to use atomic add here?

Reply:
I don't know if it is better. It is a single thread. Previously `keys` was used and it is not atomic as well.

Reply:
can  TABLE* created be created from TABLE_SHARE concurrently? ie from two connections?
Imagine two such big CTE queries from two connections? If answer is yes, I would recommend atomic

Reply:
I think we are safe here.
`add_tmp_keys()` is called only for temp tables.
As per description: `Add one key to a materialized derived table.`

The caller, `Table_ref::generate_keys()`, has assert for this at its beginning.

Reply:
Ok, if it is for temp table, then other connections/threads cannot use the same TABLE_SHARE.
I would recommend a name that indicates this variable is meant for temp-table. Something like temp_table_key_id?

Reply:
I also wonder if it is wise to add a debug assertion that validates duplicate keynames are disallowed? (they are disallowed by design for regular tables but temp tables got away with it).

My intention is that server layer catch/report this problem, not SEs.

Reply:
Thanks for the fixing the name. Optional: better to remove the server word suffix

right, I always feel the same. But there is always a new branch code that can populate TABLE* and break the assertion. But it all depends on where we do the check.

I will not insist on this. if there is a good place to check, we may do it.
@dlenev may be has some ideas/thoughts on the assertion? :thinking: